File: packages/shared-types/src/execution.ts
export type ExecutionStatus = 'in_progress' | 'completed' | 'abandoned' | 'paused';

export type StepExecutionStatus = 'pending' | 'in_progress' | 'completed' | 'skipped' | 'error';

export interface RoutineExecution {
  id: string;
  routineId: string;
  userId: string;
  status: ExecutionStatus;
  startedAt: string;
  completedAt?: string;
  pausedAt?: string;
  totalSteps: number;
  completedSteps: number;
  errorCount: number;
  stallCount: number;
  complexityAtStart: number;
  complexityAtEnd?: number;
  notes?: string;
  createdAt: string;
}

export interface StepExecution {
  id: string;
  executionId: string;
  stepId: string;
  status: StepExecutionStatus;
  startedAt?: string;
  completedAt?: string;
  durationSeconds?: number;
  errorCount: number;
  stallCount: number;
  rePromptCount: number;
  neededHelp: boolean;
  createdAt: string;
}

File: apps/neuronav_app/lib/data/models/execution.dart
import 'package:json_annotation/json_annotation.dart';

part 'execution.g.dart';

@JsonSerializable(fieldRename: FieldRename.snake)
class ExecutionRow {
  final String id;
  final String routineId;
  final String userId;
  final String status;
  final String startedAt;
  final String? completedAt;
  final int completedSteps;
  final int totalSteps;
  final int errorCount;
  final int stallCount;

  const ExecutionRow({
    required this.id,
    required this.routineId,
    required this.userId,
    required this.status,
    required this.startedAt,
    this.completedAt,
    this.completedSteps = 0,
    this.totalSteps = 0,
    this.errorCount = 0,
    this.stallCount = 0,
  });

  factory ExecutionRow.fromJson(Map<String, dynamic> json) =>
      _$ExecutionRowFromJson(json);

  Map<String, dynamic> toJson() => _$ExecutionRowToJson(this);
}

@JsonSerializable(fieldRename: FieldRename.snake)
class NewStepExecution {
  final String executionId;
  final String stepId;
  final String status;
  final int durationSeconds;
  final int errorCount;
  final int stallCount;
  final int rePromptCount;

  const NewStepExecution({
    required this.executionId,
    required this.stepId,
    required this.status,
    this.durationSeconds = 0,
    this.errorCount = 0,
    this.stallCount = 0,
    this.rePromptCount = 0,
  });

  factory NewStepExecution.fromJson(Map<String, dynamic> json) =>
      _$NewStepExecutionFromJson(json);

  Map<String, dynamic> toJson() => _$NewStepExecutionToJson(this);
}
